### ISSUE TITLE
Using Ghost Table for migration

### ISSUE THREAD

**Issue by fdrgiit** [github/gh-ost#544]
Hello Team,

I have a question based on **GH-OST** tool working.

Firstly, from the documents, and my testing I see that this is creating a separate table, applying changes to the same and then renaming this to the Original table and deleting the old table. Please correct me here, if my understanding is wrong.

So my question here are:
1. Is there a switch or command that I can use to make sure ghost-table is not dropped?
2. Is there a possibility that the created ghost table can be used to migrate it to a different database via third party tools(Eg. Attunity).
3. If there is a way to retain the **ghost table**, then if I again execute a ADD COLUMN command, can this same gh-ost table be used?

Kindly review and please share your comments. 

Thanks and Regards,
Abhinav Singh

**Comment by shlomi-noach** [comment]
> Firstly, from the documents, and my testing I see that this is creating a separate table, applying changes to the same and then renaming this to the Original table and deleting the old table. Please correct me here, if my understanding is wrong.

You are correct.

> Is there a switch or command that I can use to make sure ghost-table is not dropped?

By default, the table does _not_ get dropped. You will need a `-ok-to-drop-table`

> Is there a possibility that the created ghost table can be used to migrate it to a different database via third party tools(Eg. Attunity).

"that the created ghost table can be used to migrate it" -- migrate "it"? I'm afraid I don't follow and am not sure what it is you're trying to achieve. Please elaborate.

> If there is a way to retain the ghost table, then if I again execute a ADD COLUMN command, can this same gh-ost table be used?

Not at this time. Theoretically, this can be done if we keep track of the binlog position at the completion of the previous migration, and assuming binlogs are available. I'm wondering, why do you ask?

If you tell us the story, it will be easier to answer.

**Comment by fdrgiit** [comment]
Hello Shlomi,

Thank you very much for your prompt and detailed response.

Sir, the use-case that I have here is, my startup have customer's data stored in MySQL instances and some of the tables are really big. Hence, in-order to add columns we are using the **GH-OST** tool to make sure, we face zero downtime on the tables getting locked, when we issue an **ALTER ADD COLUMN**

At the same time, we want to migrate databases from MySQL to Oracle, for analytical purposes and for that we contemplating third part migration tools, For Eg. Attunity. But sir, during our POC, we faced an issue that, since we are using GH-OST to do this kind of schema change, the third party migration tools are not able to detect this change. The sole reason that this is not happening is because, as soon we run the following command:

`./gh-ost --host=mysqltest.caqsuj9suw0z.us-east-1.rds.amazonaws.com --user=abhinsi --password=25894237 --database=test --table=ghost_test --ok-to-drop-table --initially-drop-ghost-table --alter="ADD COLUMN LINKEDIN INT NOT NULL DEFAULT '0'" --chunk-size=2000 --max-load=Threads_connected=20 --allow-on-master --verbose --cut-over=default --exact-rowcount --concurrent-rowcount --default-retries=120 --panic-flag-file=/root/ghost.panic.flag --postpone-cut-over-flag-file=/root/ghost.postpone.flag --execute`

The tool is creating three tables namely:

**Creating changelog table `test`.`_ghost_test_ghc`
Creating ghost table `test`.`_ghost_test_gho`
Creating magic cut-over table `test`.`_ghost_test_del`**

And the changes are happening on these tables and then at last a **RENAME** is being on the GH-OST table. 

So sir, what I was thinking is if there is a possibility to create this GH-OST table only once and then whatever changes happens on the parent tables are being moved to this GH-OST table, so that we can use the migration task to point to this GH-OST table and be updated.

Lastly sir, the `-ok-to-drop-table` switch, still drops the **ghc** and **gho** but leaves the **del**. Please correct me if I am wrong on this part.

Kindly review and please share your comments.

**Comment by shlomi-noach** [comment]
Thank you for elaborating, it's easier now to respond and it is an interesting use case.

> since we are using GH-OST to do this kind of schema change, the third party migration tools are not able to detect this change. The sole reason that this is not happening is because, as soon we run the following command ... The tool is creating three tables ... And the changes are happening on these tables and then at last a RENAME is being on the GH-OST table.

This confuses me, because the changes are still happening on your _original_ table, where your tools, such as Attunity, should be able to observe.

To explain a little bit more in depth:

- `_ghost_test_ghc` is of no interest to the user; it is used for internal coordination between operations on the master and on the replica. You certainly don't need Attunity to be aware of the table.
- `_ghost_test_del` is only created at the very end of the operation, and once created, turns irrelevant to you. So, again, Attunity shouldn't care about this table.
- `_ghost_test_gho` is naturally the most interesting table. But do **note** that every change on this table reflects a change on the _original_ table. If you continue to follow your _original_ table you should _keep on seeing the real changes_. 
- I have no idea whether Attunity captures the `RENAME` itself. Actually, the ideal situation for you in this case is that it _doesn't_. It should see the _original_ table with ongoing changes throughout the operation, and then suddenly it should see what it _thinks_ is the _original_ table, but is in face the renamed `_gho` table, with new incoming changes.
- I can only think of a single use case which might be a concern for you, but I'm unsure whether it is a use case you've seen: adding a column with a non-default `DEFAULT` value. Like add an `INT DEFAULT 17`.

What I'm missing from your description is where and when exactly the failure happens. _What went wrong?_

> So sir, what I was thinking is if there is a possibility to create this GH-OST table only once and then whatever changes happens on the parent tables are being moved to this GH-OST table, so that we can use the migration task to point to this GH-OST table and be updated.

This can't be done, for it requires infinite capture of binary logs. If I run two migrations on a table, three months apart, that means I need to capture three months of binary logs.

Is is possible to solve this _ad-hoc_ but not in the general case.

> Lastly sir, the -ok-to-drop-table switch, still drops the ghc and gho but leaves the del. Please correct me if I am wrong on this part.

According to the following, the `_del` table is dropped when `-ok-to-drop-table` is specified:

https://github.com/github/gh-ost/blob/ce4e42da328ddad57ddea204fd768b91e58bdd96/go/logic/migrator.go#L1229-L1232

**Comment by fdrgiit** [comment]
Hello Sir, 

Thank you very much for your prompt and detailed response.

Yes sir, you got it perfectly right that the **RENAME** operation is not being captured by the Attunity tool. Hence, the reason sir for me to reachout to you. Since, `RENAME `is a documented limitation here, the **ALTER ADD COLUMN** operation via GH-OST never makes it to the target table.

Sir, if you can think of any other tools other than GH-OST for online schema, that can help me in this scenario, please let me know. Or if you have any other suggestions, please comment. This is really important for us.

Thanks and Regards,
Abhinav Singh

**Comment by shlomi-noach** [comment]
All online migration tools behave similarly with RENAME, unfortunately...

**Comment by fdrgiit** [comment]
Hello @shlomi-noach ,

Thank you very much for the information.

One thing that I noticed here, when I used the following command:

`./gh-ost --host=mysqltest.caqsuj9suw0z.us-east-1.rds.amazonaws.com --user=abhinsi --password=25894237 --database=test --table=ghost_test --ok-to-drop-table --initially-drop-ghost-table --alter="ADD COLUMN LINKEDIN INT NOT NULL DEFAULT '0'" --chunk-size=2000 --max-load=Threads_connected=20 --allow-on-master --verbose --cut-over=default --exact-rowcount --concurrent-rowcount --default-retries=120 --panic-flag-file=/root/ghost.panic.flag --postpone-cut-over-flag-file=/root/ghost.postpone.flag --execute`

I could see the following logs:

> 2018-02-08 17:15:41 INFO Ghost table created
> 2018-02-08 17:15:41 INFO Altering ghost table `test`.`_ghost_test_gho`
> 2018-02-08 17:15:41 INFO Ghost table altered
> 2018-02-08 17:15:41 INFO Intercepted changelog state GhostTableMigrated
> 2018-02-08 17:15:41 INFO Waiting for ghost table to be migrated. Current lag is 0s
> 2018-02-08 17:15:41 INFO Handled changelog state GhostTableMigrated
> 2018-02-08 17:15:41 INFO Chosen shared unique key is PRIMARY
> 2018-02-08 17:15:41 INFO Shared columns are ID,SCHED_NAME,ENTRY_ID,TRIGGER_NAME,TRIGGER_GROUP,INSTANCE_NAME,FIRED_TIME,SCHED_TIME,PRIORITY,STATE,JOB_NAME,JOB_GROUP,IS_NONCONCURRENT,REQUESTS_RECOVERY,GMAIL,GHOSTTING,TWEET,TWITTER,LINKEDIN
> 2018-02-08 17:15:41 INFO Listening on unix socket file: /tmp/gh-ost.test.ghost_test.sock
> 2018-02-08 17:15:41 INFO As instructed, counting rows in the background; meanwhile I will use an estimated count, and will update it later on
> 2018-02-08 17:15:41 INFO As instructed, I'm issuing a SELECT COUNT(*) on the table. This may take a while
> 2018-02-08 17:15:41 INFO Migration min values: [2]
> 2018-02-08 17:15:41 INFO Exact number of rows via COUNT: 1099
> 2018-02-08 17:15:41 INFO Migration max values: [1100]
> 2018-02-08 17:15:41 INFO Waiting for first throttle metrics to be collected
> 2018-02-08 17:15:41 INFO First throttle metrics collected
> Migrating `test`.`ghost_test`; Ghost table is `test`.`_ghost_test_gho`
> Migrating ip-10-1-0-90:3306; inspecting ip-10-1-0-90:3306; executing on ip-172-31-69-254
> Migration started at Thu Feb 08 17:15:41 +0100 2018
> chunk-size: 2000; max-lag-millis: 1500ms; dml-batch-size: 10; max-load: Threads_connected=20; critical-load: ; nice-ratio: 0.000000
> throttle-additional-flag-file: /tmp/gh-ost.throttle
> postpone-cut-over-flag-file: /root/ghost.postpone.flag
> panic-flag-file: /root/ghost.panic.flag
> Serving on unix socket: /tmp/gh-ost.test.ghost_test.sock
> Copy: 0/1099 0.0%; Applied: 0; Backlog: 0/1000; Time: 0s(total), 0s(copy); streamer: mysql-bin-changelog.085234:2121; State: migrating; ETA: N/A
> Copy: 0/1099 0.0%; Applied: 0; Backlog: 0/1000; Time: 1s(total), 1s(copy); streamer: mysql-bin-changelog.085234:5644; State: migrating; ETA: N/A
> 2018-02-08 17:15:42 INFO Row copy complete
> Copy: 1099/1099 100.0%; Applied: 0; Backlog: 0/1000; Time: 1s(total), 1s(copy); streamer: mysql-bin-changelog.085234:96313; State: migrating; ETA: due
> 2018-02-08 17:15:42 INFO Grabbing voluntary lock: gh-ost.15504.lock

As you can see that GH-OST tool, created the **gho** table and then did the **ALTER** to add the column and then did the migration. So sir, what I did is in Attunity, I tried migrating the **gho** table. But sir, the issue here is that this table get's picked up by the migration table and even the ALTER operation to ADD the column, but the migration part of moving data from the parent table to the **gho** table seems to me is not a logged operation in BINARY LOGS.

So sir, what I am asking is, there any possibility that I can make this logged operation. That would atleast help a certain bit here.

Kindly review and please share your comments.

**Comment by shlomi-noach** [comment]
OK, I understand now. Attunity misses the `ALTER` statement.

> the migration part of moving data from the parent table to the gho table seems to me is not a logged operation in BINARY LOGS.

It is absolutely logged in the binary logs. The entire operation depends on it.

**Comment by fdrgiit** [comment]
Hello @shlomi-noach 

> It is absolutely logged in the binary logs. The entire operation depends on it.

What I am trying to ask here is, the part where the parent table have let's say 30 records already and then you use **GH-OST** for doing online schema change and the part where the the ghost table is created, can you please tell me how are these 30 records being migrated and are the related transactions for migrating these 30 records, again being logged in the Binary Logs.

Since sir, if the initial data migration is a logged migration, then I can try to use the feature to migrate the ghost table rather than the parent table.

I hope I am making some sense here. Kindly review and please share your comments.

Thanks and Regards,
Abhinav Singh

**Comment by shlomi-noach** [comment]
the _ghost_ table is populated by either:

- `insert into <ghots table> (...) select ... from <original table>`, or:
- `insert`, `update`, `delete` directly on the ghost table.

To elaborate more on that would be to explain the entire logic of gh-ost, and description of that is found in the documentation, hopefully.

Again, the changes to the _ghost_ table _are logged_.
